File: sdks/node/src/workflows.ts
import type { ApplAdServer } from './client';

export class Workflows {
  constructor(private client: ApplAdServer) {}

  create(
    name: string,
    opts?: {
      description?: string;
      triggerType?: string;
      triggerConfig?: Record<string, unknown>;
      nodes?: unknown[];
      edges?: unknown[];
    }
  ) {
    return this.client.call('POST', '/workflows', {
      name,
      description: opts?.description ?? '',
      triggerType: opts?.triggerType ?? 'manual',
      triggerConfig: opts?.triggerConfig ?? {},
      nodes: opts?.nodes ?? [],
      edges: opts?.edges ?? [],
    });
  }

  list() {
    return this.client.call('GET', '/workflows');
  }

  get(workflowId: string) {
    return this.client.call('GET', `/workflows/${workflowId}`);
  }

  update(workflowId: string, data: Record<string, unknown>) {
    return this.client.call('PUT', `/workflows/${workflowId}`, data);
  }

  delete(workflowId: string) {
    return this.client.call('DELETE', `/workflows/${workflowId}`);
  }

  execute(workflowId: string, triggerData?: Record<string, unknown>) {
    return this.client.call('POST', `/workflows/${workflowId}/execute`, {
      triggerData: triggerData ?? {},
    });
  }

  listExecutions(workflowId: string) {
    return this.client.call('GET', `/workflows/${workflowId}/executions`);
  }

  getExecution(workflowId: string, executionId: string) {
    return this.client.call('GET', `/workflows/${workflowId}/executions/${executionId}`);
  }
}

File: sdks/dart/lib/src/workflows.dart
import 'package:dio/dio.dart';

/// Workflows service — manage workflow definitions and executions.
class Workflows {
  final Dio _dio;

  Workflows(this._dio);

  /// Create a workflow.
  Future<Map<String, dynamic>> create({
    required String name,
    String? description,
    String triggerType = 'manual',
    Map<String, dynamic>? triggerConfig,
    List<Map<String, dynamic>>? nodes,
    List<Map<String, dynamic>>? edges,
  }) async {
    final res = await _dio.post('/v1/workflows', data: {
      'name': name,
      if (description != null) 'description': description,
      'triggerType': triggerType,
      if (triggerConfig != null) 'triggerConfig': triggerConfig,
      if (nodes != null) 'nodes': nodes,
      if (edges != null) 'edges': edges,
    });
    return res.data;
  }

  /// List workflows.
  Future<Map<String, dynamic>> list() async {
    final res = await _dio.get('/v1/workflows');
    return res.data;
  }

  /// Get a workflow by ID.
  Future<Map<String, dynamic>> get(String workflowId) async {
    final res = await _dio.get('/v1/workflows/$workflowId');
    return res.data;
  }

  /// Update a workflow.
  Future<Map<String, dynamic>> update(
    String workflowId,
    Map<String, dynamic> data,
  ) async {
    final res = await _dio.put('/v1/workflows/$workflowId', data: data);
    return res.data;
  }

  /// Delete a workflow.
  Future<void> delete(String workflowId) async {
    await _dio.delete('/v1/workflows/$workflowId');
  }

  /// Execute a workflow.
  Future<Map<String, dynamic>> execute(
    String workflowId, {
    Map<String, dynamic>? triggerData,
  }) async {
    final res = await _dio.post('/v1/workflows/$workflowId/execute', data: {
      if (triggerData != null) 'triggerData': triggerData,
    });
    return res.data;
  }

  /// List executions for a workflow.
  Future<Map<String, dynamic>> listExecutions(String workflowId) async {
    final res = await _dio.get('/v1/workflows/$workflowId/executions');
    return res.data;
  }

  /// Get a single execution.
  Future<Map<String, dynamic>> getExecution(
    String workflowId,
    String executionId,
  ) async {
    final res = await _dio
        .get('/v1/workflows/$workflowId/executions/$executionId');
    return res.data;
  }
}
